api/python/t4/api.py

```diff
@@ -208,23 +208,188 @@ def list_packages(registry=None):
+        pkg_namespaces = [result['Prefix'] for result in pkg_namespaces]
+
+        # go through package namespaces to get packages
+        for pkg_namespace in pkg_namespaces:
+            pkg_names, _ = list_objects(
+                bucket_name,
+                pkg_namespace,
+                recursive=False
+            )
+            pkg_names = [pkg_name['Prefix'] for pkg_name in pkg_names]
+
+            # go through packages to get package hash files
+            for pkg_name in pkg_names:
+                pkg_hashes = []
+                pkg_sizes = []
+                pkg_ctimes = []
+                pkg_names = []
+
+                _, pkg_hashfiles = list_objects(
+                    bucket_name,
+                    pkg_name,
+                    recursive=False
+                )
+
+                latest_hashfile = next(hf for hf in pkg_hashfiles if '/latest' in hf['Key'])['Key']
+                latest_hashfile_fullpath = f's3://{bucket_name}/{latest_hashfile}'
+                latest_hash, _ = get_bytes(latest_hashfile_fullpath)
+                latest_hash = latest_hash.decode()
+
+                # go through package hashfiles to get manifest files
+                pkg_hash_paths = []
+                for pkg_hashfile in pkg_hashfiles:
+                    pkg_hashfile_key = pkg_hashfile['Key']
+
+                    if pkg_hashfile_key.split('/')[-1] == 'latest':
+                        continue
+
+                    pkg_hash, _ = get_bytes(f's3://{bucket_name}/{pkg_hashfile_key}')
+                    pkg_hash = pkg_hash.decode()
+
+                    pkg_last_modified = pkg_hashfile['LastModified'].timestamp()
+                    pkg_hash_paths.append(pkg_hashfile_key)
+                    pkg_ctimes.append(pkg_last_modified)
+
+                # go through manifest files to get package info
+                for pkg_hash_path in pkg_hash_paths:
+                    name = pkg_name[len(bucket_registry_path):].strip('/')
+
+                    pkg_hash, _ = get_bytes('s3://' + bucket_name + '/' + pkg_hash_path)
+                    pkg_hash = pkg_hash.decode()
+
+                    if pkg_hash == latest_hash:
+                        pkg_name = name + ':latest'
+                    else:
+                        pkg_name = name
+
+                    pkg_names.append(pkg_name)
+                    pkg_hashes.append(pkg_hash)
+
+                    from t4 import Package
+                    pkg = Package.browse(
+                        pkg_name, pkg_hash=pkg_hash, registry='s3://' + bucket_name
+                    )
+                    pkg_sizes.append(pkg.reduce(lambda tot, tup: tot + tup[1].size, default=0))
+
+                pkg_info += [(pkg_name, hash, ctime, size) for (pkg_name, hash, ctime, size) in
+                             zip(pkg_names, pkg_hashes, pkg_ctimes, pkg_sizes)]
+
+        return PackageList(pkg_info)
```

Comment:
Probably need to sort it here for consistency too.

Might as well drop it out of the if statement too.
```
if(file):
    ...
elif(s3):
    ...
else:
    raise

def Sorter():
   ...
pkg_info = sorted(pkg_info, key=sorter)
return PackageList(pkg_info)
```